prac_07/project_management.py

```diff
@@ -0,0 +1,183 @@
+            projects.append(project)
+    return projects
+
+
+def save_projects(project_filename, projects):
+    """Load projects to a file from a list of projects."""
+    with open(project_filename, 'w', newline='', encoding='utf-8') as out_file:
+        writer = csv.writer(out_file, delimiter='\t')
+        header = ["Name", "Start Date", "Priority", "Cost Estimate", "Completion Percentage"]
+        writer.writerow(header)
+        for project in projects:
+            row = [project.name, project.start_date, project.priority,
+                   project.cost_estimate, project.completion_percentage]
+            writer.writerow(row)
+
+
+def display_projects(projects):
+    """Display projects and whether they are complete."""
+    incomplete_projects = [project for project in projects if not project.is_completed()]
+    print("Incomplete Projects:")
+    for project in incomplete_projects:
+        print(f"  {project}")
+    print("Completed Projects:")
+    completed_projects = [project for project in projects if project.is_completed()]
+    for project in completed_projects:
+        print(f"  {project}")
+
+
+def filter_by_date(projects, date_string):
+    """Display projects that start after a certain date."""
+    date = datetime.datetime.strptime(date_string, "%d/%m/%Y").date()
+    for project in projects:
+        if project.start_date > date:
+            print(project)
+
+
+def add_project(projects):
+    """Add a valid project from user input."""
+    print("Let's add a new project")
+    name = input("Name: ")
+    is_valid_date = False
+    is_valid_priority = False
+    is_valid_cost = False
+    is_valid_completion = False
+    while not is_valid_date:
+        try:
+            date_string = input("Start date (dd/mm/yyyy): ")
+            date = datetime.datetime.strptime(date_string, "%d/%mm/%Y").date()
+            is_valid_date = True
+        except ValueError:
+            print("Enter valid date.")
+    while not is_valid_priority:
+        try:
+            priority = int(input("Priority: "))
+            if priority >= 1:
+                is_valid_priority = True
+            else:
+                print("Enter valid priority.")
+        except TypeError:
+            print("Enter valid priority.")
+    while not is_valid_cost:
+        try:
+            cost_estimate = float(input("Cost estimate: $"))
+            if cost_estimate >= 0:
+                is_valid_cost = True
+            else:
+                print("Enter valid cost estimate.")
+        except TypeError:
+            print("Enter valid cost estimate.")
+    while not is_valid_completion:
+        try:
+            completion_percentage = int(input("Percent complete: "))
+            if 0 < completion_percentage <= 100:
+                is_valid_completion = True
+            else:
+                print("Enter valid completion percent.")
+        except TypeError:
+            print("Enter valid completion percent.")
```

Comment:
You could make all four of these error checking procedures into a single function